### PR TITLE
Mixed dedup and compression

### PR DESCRIPTION
**Pull request by bardavid**
Introducing support for dedupe and compression
on the same job. When used together, compression is
calculated from unique capacity. E.g. when using
dedupe_percentage=50 and buffer_compress_percentage=50,
then total reduction should be 75% - 50% would be deduped
while 50% of the remaining buffers would be compressed

### Motivation
position fio as a performance benchmarking tool for general purpose storage providers.

### Example runs:

50/50 dedup compression. with these settings, per 100mb written we should expect 25mb physical space consumption if the underlying storage device compresses and dedupes.. i.e. DRR 4:1. 
`$ ./fio --direct=1 --dedupe_percentage=50 --name=bar3 --bs=256k --iodepth=10 --size=100m  --rw=write --buffer_compress_percentage=50 --ioengine=libaio --dedupe_mode=working_set`

and to test results with the adjusted fio-dedup binary:
```
$ ./t/fio-dedupe -b 262144 -C 1 bar3.0.0 

Will check <bar3.0.0>, size <104857600>, using 15 threads
Threads(15): 400 items processed
Extents=400, Unique extents=195 Duplicated extents=20
De-dupe ratio: 1:1.05
De-dupe working set at least: 5.00%
Fio setting: dedupe_percentage=51
Unique capacity 24MB
```

with dedupe_mode=repeat and smaller blocksize:
`$ ./fio --direct=1 --dedupe_percentage=50 --name=bar4 --bs=8k --iodepth=10 --size=100m  --rw=write --buffer_compress_percentage=50 --ioengine=libaio`

testing the results with the binary, note that it finds a very large (25%) dedupe working set since we are repeating previous buffers when generating the ddp buffers.
```
$ ./t/fio-dedupe -b 8192 -C 1 bar4.0.0 
Will check <bar4.0.0>, size <104857600>, using 15 threads
Threads(15): 12800 items processed
Extents=12800, Unique extents=6356 Duplicated extents=3167
De-dupe ratio: 1:1.01
De-dupe working set at least: 24.74%
Fio setting: dedupe_percentage=50
Unique capacity 24MB
```

with 75% compression
`$ ./fio --direct=1 --dedupe_percentage=50 --name=bar5 --bs=8k --iodepth=10 --size=100m  --rw=write --buffer_compress_percentage=75 --ioengine=libaio --dedupe_mode=working_set`

```
$ ./t/fio-dedupe -b 8192 -C 1 bar5.0.0 
Will check <bar5.0.0>, size <104857600>, using 15 threads
Threads(15): 12800 items processed
Extents=12800, Unique extents=6356 Duplicated extents=640
De-dupe ratio: 1:1.01
De-dupe working set at least: 5.00%
Fio setting: dedupe_percentage=50
Unique capacity 13MB
```

### CODE REVIEW

**Comment by bardavid** [comment]
@sitsofe wasn't sure who to approach with promoting this, can you please assist ? thanks !

File: t/dedupe.c [hunk 1]
```diff
@@ -304,13 +347,29 @@ static int do_work(struct worker_thread *thread, void *buf)
...
+		return;
+
+	thread->zc.buf_in = fio_memalign(blocksize, blocksize, false);
+	thread->zc.buf_out = fio_memalign(blocksize, deflateBound(stream, blocksize), false);
+}
+
 static void *thread_fn(void *data)
 {
-	struct worker_thread *thread = data;
+	struct worker_thread *thread = my_wt = data;
```
Comment (axboe) [inline t/dedupe.c:366]: Please don't assign the thread variable together with the local one, do that separately.

**Comment by axboe** [comment]
First commit looks reasonable, but the second one needs some editing.

File: t/dedupe.c (hunk 1 above)
Comment (bardavid) [reply t/dedupe.c]: fixed

File: t/dedupe.c [hunk 2]
```diff
@@ -495,12 +581,16 @@ static void show_stat(uint64_t nextents, uint64_t nchunks, uint64_t ndupextents)
 	perc *= 100.0;
 	printf("Fio setting: dedupe_percentage=%u\n", (int) (perc + 0.50));
 
+
+	if (compression) {
+	    bytes_to_human_readable_unit(unique_capacity, &uc_human, &unit);
+	    printf("Unique capacity %lu%s\n", uc_human, unit);
+	}
```
Comment (ammarfaizi2) [inline t/dedupe.c:588]: Inconsistent indentation here.
```suggestion
	if (compression) {
		bytes_to_human_readable_unit(unique_capacity, &uc_human, &unit);
		printf("Unique capacity %lu%s\n", uc_human, unit);
	}
```

File: t/dedupe.c [hunk 3]
```diff
@@ -473,9 +535,33 @@ static void show_chunk(struct chunk *c)
...
+
+out:
+    *unit_out = capacity_unit[i];
+}
+
+static void show_stat(uint64_t nextents, uint64_t nchunks, uint64_t ndupextents, uint64_t unique_capacity)
 {
 	double perc, ratio;
+    const char *unit;
+    uint64_t uc_human;
```
Comment (ammarfaizi2) [inline t/dedupe.c:564]: Inconsistent indentation here.
```suggestion
	double perc, ratio;
	const char *unit;
	uint64_t uc_human;
```

File: t/dedupe.c [hunk 4]
```diff
@@ -24,6 +24,13 @@
 
 #include "../lib/bloom.h"
 #include "debug.h"
+#include "zlib.h"
+
+struct zlib_ctrl {
+    z_stream stream;
+    unsigned char *buf_in;
+    unsigned char *buf_out;
+};
```
Comment (ammarfaizi2) [inline t/dedupe.c:33]: Please use tabs.
```suggestion
struct zlib_ctrl {
	z_stream stream;
	unsigned char *buf_in;
	unsigned char *buf_out;
};
```

File: t/dedupe.c [hunk 5]
```diff
@@ -37,6 +44,9 @@ struct worker_thread {
 	unsigned long items;
 	unsigned long dupes;
 	int err;
+
+    unsigned long long unique_capacity;
+    struct zlib_ctrl zc;
```
Comment (ammarfaizi2) [inline t/dedupe.c:49]: Inconsistent indentation here...
```suggestion
	unsigned long items;
	unsigned long dupes;
	int err;

 	unsigned long long unique_capacity;
 	struct zlib_ctrl zc;
```

File: t/dedupe.c [hunk 6]
```diff
@@ -473,9 +535,33 @@ static void show_chunk(struct chunk *c)
...
+    }
+
+    while (*n_out >= 1024) {
+        i++;
+        *n_out /= 1024;
+    }
+
+out:
+    *unit_out = capacity_unit[i];
+}
```
Comment (ammarfaizi2) [inline t/dedupe.c:558]: Use tabs. Also single statement inside "if" doesn't have to be closed with braces.

```suggestion
static void bytes_to_human_readable_unit(uint64_t n, uint64_t *n_out, const char **unit_out)
{
	uint8_t i;

	*n_out = n;
	i = 0;

	if (*n_out < 1024)
		goto out;

	while (*n_out >= 1024) {
		i++;
		*n_out /= 1024;
	}

out:
	*unit_out = capacity_unit[i];
}
```

**Review by ammarfaizi2 — CHANGES_REQUESTED** [review]
Hi,
Small trivial code style issues.

File: t/dedupe.c (hunk 2 above)
Comment (bardavid) [reply t/dedupe.c]: fixed

File: t/dedupe.c (hunk 3 above)
Comment (bardavid) [reply t/dedupe.c]: fixed

File: t/dedupe.c (hunk 4 above)
Comment (bardavid) [reply t/dedupe.c]: fixed

File: t/dedupe.c (hunk 5 above)
Comment (bardavid) [reply t/dedupe.c]: fixed

File: t/dedupe.c (hunk 6 above)
Comment (bardavid) [reply t/dedupe.c]: fixed

**Comment by bardavid** [comment]
ready for re-review ..

**Comment by axboe** [comment]
There are still unadressed review comments from the previous round, please address those first.

**Comment by axboe** [comment]
Not as far as I can tell, in fact when I click the commit they are still listed as pending. I had three comments in there last time, just commented on all 3 again.

**Comment by bardavid** [comment]
> Not as far as I can tell, in fact when I click the commit they are still listed as pending. I had three comments in there last time, just commented on all 3 again.

Have you submitted those comments yet ? I can't see them.

**Comment by axboe** [comment]
Just click the commit, they are in there.

**Comment by ammarfaizi2** [comment]
> Just click the commit, they are in there.

I don't see that. Have you clicked "Submit Review" button Jens?

**Comment by axboe** [comment]
Looks like I didn't, I wonder if github changed single review comments to now require submit, before they just submitted. Or maybe I didn't do single comments this time and messed up, not sure. Should be there now.

**Comment by bardavid** [comment]
fixed per comments @axboe

File: t/dedupe.c [hunk 7]
```diff
@@ -495,12 +569,16 @@ static void show_stat(uint64_t nextents, uint64_t nchunks, uint64_t ndupextents)
 	perc *= 100.0;
 	printf("Fio setting: dedupe_percentage=%u\n", (int) (perc + 0.50));
 
+
+	if (compression) {
+		uc_human = bytes_to_human_readable_unit(unique_capacity, &unit);
+		printf("Unique capacity %lu%s\n", uc_human, unit);
```
Comment (ammarfaizi2) [inline t/dedupe.c:575]: Got a warning here:
```
t/dedupe.c:575:43: warning: format ‘%lu’ expects argument of type ‘long unsigned int’, but argument 2 has type ‘uint64_t’ {aka ‘long long unsigned int’} [-Wformat=]
  575 |                 printf("Unique capacity %lu%s\n", uc_human, unit);
      |                                         ~~^       ~~~~~~~~
      |                                           |       |
      |                                           |       uint64_t {aka long long unsigned int}
      |                                           long unsigned int
      |                                         %llu
```

The specifier for `uint64_t` may not always be (unsigned long). It is sometimes (unsigned long long).

So please cast this one to `unsigned long` like the other parts.
```suggestion
		printf("Unique capacity %lu%s\n", (unsigned long) uc_human, unit);
```

**Review by ammarfaizi2 — CHANGES_REQUESTED** [review]

File: t/dedupe.c [hunk 8]
```diff
@@ -135,6 +142,33 @@ static int read_block(int fd, void *buf, off_t offset)
...
+	stream->avail_out = deflateBound(stream, blocksize);
+	stream->next_out = zc->buf_out;
+
+	ret = deflate(stream, Z_FINISH);
+	assert(ret != Z_STREAM_ERROR);
+	compressed_len = blocksize - stream->avail_out;
+
+	if (dump_output)
+		printf("offset 0x%lx compressed to %d blocksize %d ratio %.2f \n",
+				offset, compressed_len, blocksize, (float)compressed_len / (float)blocksize);
```
Comment (ammarfaizi2) [inline t/dedupe.c:165]: Got a warning here
```t/dedupe.c: In function ‘account_unique_capacity’:
t/dedupe.c:164:36: warning: format ‘%lx’ expects argument of type ‘long unsigned int’, but argument 2 has type ‘uint64_t’ {aka ‘long long unsigned int’} [-Wformat=]
  164 |                 printf("offset 0x%lx compressed to %d blocksize %d ratio %.2f \n",
      |                                  ~~^
      |                                    |
      |                                    long unsigned int
      |                                  %llx
  165 |                                 offset, compressed_len, blocksize, (float)compressed_len / (float)blocksize);
      |                                 ~~~~~~
      |                                 |
      |                                 uint64_t {aka long long unsigned int}
```
```suggestion

		printf("offset 0x%lx compressed to %d blocksize %d ratio %.2f \n",
		       (unsigned long) offset, compressed_len, blocksize, 
		       (float)compressed_len / (float)blocksize);
```

File: t/dedupe.c (hunk 7 above)
Comment (bardavid) [reply t/dedupe.c:575]: fixed

File: t/dedupe.c (hunk 8 above)
Comment (bardavid) [reply t/dedupe.c:165]: fixed

**Comment by bardavid** [comment]
ready for re-review

**Pull request merged**